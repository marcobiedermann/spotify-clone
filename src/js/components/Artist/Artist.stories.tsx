import type { Meta, StoryObj } from '@storybook/react';
import Artist from '.';
import artistFixtures from './__fixtures__';

const meta: Meta<typeof Artist> = {
  component: Artist,
  args: artistFixtures,
};

type Story = StoryObj<typeof Artist>;

const Default: Story = {};

export { Default };
export default meta;
