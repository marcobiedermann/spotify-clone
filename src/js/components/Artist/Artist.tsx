import React from 'react';
import Image from '../Image';

const defaultImageSize = 100;

interface Image {
  height: number | null;
  url: string;
  width: number | null;
}

export interface ArtistProps {
  name?: string;
  id?: string;
  images?: Image[];
}

function Artist(props: ArtistProps): JSX.Element {
  const { images = [], name } = props;
  const image = images[0];

  return (
    <figure>
      {image && (
        <Image
          {...image}
          alt={name || ''}
          width={image.width || defaultImageSize}
          height={image.height || defaultImageSize}
        />
      )}
      <figcaption>
        <h2>{name}</h2>
      </figcaption>
    </figure>
  );
}

export default Artist;
