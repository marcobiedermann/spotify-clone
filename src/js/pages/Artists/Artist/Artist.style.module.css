.artist__image {
  border-radius: 50%;
}
