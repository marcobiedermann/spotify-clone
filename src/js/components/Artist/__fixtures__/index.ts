import { ArtistProps } from '../Artist';
import imageFixtures from '../../Image/__fixtures__';

const artist: ArtistProps = {
  id: 'string',
  images: [imageFixtures],
  name: 'string',
};

export default artist;
