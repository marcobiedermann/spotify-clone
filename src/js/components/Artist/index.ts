import Artist from './Artist';

export default Artist;
export * from './Artist';
