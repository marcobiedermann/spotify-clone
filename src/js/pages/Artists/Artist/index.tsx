import React from 'react';
import { Helmet } from 'react-helmet';
import { Link, RouteChildrenProps, useLocation, useParams } from 'react-router-dom';
import useSWR from 'swr';
import Artist from '../../../components/Artist';
import Error from '../../../components/Error';
import Loader from '../../../components/Loader';

interface Params {
  artistId: string;
}

interface ExternalUrls {
  spotify: string;
}

interface Followers {
  href: null;
  total: number;
}

interface Image {
  height: number;
  url: string;
  width: number;
}

export interface ArtistData {
  external_urls: ExternalUrls;
  followers: Followers;
  genres: string[];
  href: string;
  id: string;
  images: Image[];
  name: string;
  popularity: number;
  type: string;
  uri: string;
}

export type ArtistPageProps = RouteChildrenProps;

function ArtistPage(): JSX.Element {
  const { pathname } = useLocation();
  const { artistId } = useParams<Params>();
  const { data, error } = useSWR<ArtistData>(`/v1/artists/${artistId}`);

  if (error) {
    return <Error>{error.message}</Error>;
  }

  if (!data) {
    return <Loader />;
  }

  const { name } = data;

  return (
    <>
      <Helmet>
        <title>{name}</title>
      </Helmet>
      <Artist {...data} />
      <ul>
        <li>
          <Link to={`${pathname}/albums`}>Albums</Link>
        </li>
        <li>
          <Link to={`${pathname}/related-artists`}>Related Artists</Link>
        </li>
        <li>
          <Link to={`${pathname}/top-tracks`}>Top Tracks</Link>
        </li>
      </ul>
    </>
  );
}

export default ArtistPage;
