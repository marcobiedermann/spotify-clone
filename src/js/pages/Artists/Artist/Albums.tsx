/* eslint-disable jsx-a11y/label-has-associated-control */

import React from 'react';
import { useForm } from 'react-hook-form';
import { Link, useLocation, useParams } from 'react-router-dom';
import { useArtistAlbums } from '../../../hooks/artists';
import { Albums, Error, Loader, Section } from '../../../components';

interface FormData {
  include_group: 'album' | 'single';
}

function Discography() {
  const { register, watch } = useForm<FormData>({
    defaultValues: {
      include_group: 'album',
    },
  });
  const include_group = watch('include_group');
  const { pathname } = useLocation();
  const { artistId } = useParams();
  const { data, error, isError, isPending } = useArtistAlbums(artistId!, {
    include_groups: [include_group],
    limit: 9,
  });

  if (isError) {
    return <Error>{error.message}</Error>;
  }

  if (isPending) {
    return <Loader />;
  }

  const { items } = data;

  return (
    <Section>
      <div
        style={{
          alignItems: 'center',
          display: 'flex',
          gap: '1em',
          justifyContent: 'space-between',
        }}
      >
        <h2>Discography</h2>
        <p>
          <Link to={`${pathname}/albums`}>Show all</Link>
        </p>
      </div>
      <form>
        <div>
          <label htmlFor="album">Album</label>
          <input id="album" type="radio" value="album" {...register('include_group')} />
        </div>
        <div>
          <label htmlFor="single">Single</label>
          <input id="single" type="radio" value="single" {...register('include_group')} />
        </div>
      </form>
      <Albums items={items} />
    </Section>
  );
}

export default Discography;
