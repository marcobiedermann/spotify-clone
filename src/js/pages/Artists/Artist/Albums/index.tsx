import React from 'react';
import { Helmet } from 'react-helmet';
import { useParams } from 'react-router-dom';
import { Albums, Error, Loader } from '../../../../components';
import { useArtistAlbums } from '../../../../hooks/artists';

function AlbumsPage(): JSX.Element {
  const { artistId } = useParams();
  const { data, error, isError, isPending } = useArtistAlbums(artistId!);

  if (isError) {
    return <Error>{error.message}</Error>;
  }

  if (isPending) {
    return <Loader />;
  }

  const { items } = data;

  return (
    <>
      <Helmet>
        <title>Albums</title>
      </Helmet>
      <Albums items={items} />
    </>
  );
}

export default AlbumsPage;
