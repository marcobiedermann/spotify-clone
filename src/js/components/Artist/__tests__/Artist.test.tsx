import { render } from '@testing-library/react';
import React from 'react';
import Artist from '..';
import artistFixtures from '../__fixtures__';

describe('Artist component', () => {
  it('renders correctly', () => {
    const { container } = render(<Artist {...artistFixtures} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
