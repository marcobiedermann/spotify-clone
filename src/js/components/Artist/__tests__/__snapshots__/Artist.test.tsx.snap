// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Artist component > renders correctly 1`] = `
<figure>
  <img
    alt="string"
    height="300"
    src="https://i.scdn.co/image/ab67616d00001e02ff9ca10b55ce82ae553c8228"
    width="300"
  />
  <figcaption>
    <h2>
      string
    </h2>
  </figcaption>
</figure>
`;
